examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,284 @@
+
+        # initialize the weights
+        self.resnet.apply(self.init_weights)
+
+    def init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            torch.nn.init.xavier_uniform_(m.weight)
+            m.bias.data.fill_(0.01)
+
+    def forward_once(self, x):
+        output = self.resnet(x)
+        output = output.view(output.size()[0], -1)
+        return output
+
+    def forward(self, input):
+
+        # pass the input through resnet
+        output = self.forward_once(input)
+
+        # pass the output of resnet to sigmoid layer
+        output = self.sigmoid(output)
+
+        return output
+
+
+class MatcherDataset(Dataset):
+    # following class implements data downloading and handles preprocessing
+    def __init__(self, root, train, download=False):
+        super(MatcherDataset, self).__init__()
+
+        # get CIFAR10 dataset
+        self.dataset = datasets.CIFAR10(root, train=train, download=download)
+
+        # convert data from numpy array to Tensor
+        self.data = torch.from_numpy(self.dataset.data)
+
+        # shift the dimensions of dataset to match the initial input layer dimensions
+        self.data = torch.movedim(self.data, (0, 1, 2, 3), (0, 2, 3, 1))
+
+        # convert targets list to torch Tensor
+        self.dataset.targets = torch.Tensor(self.dataset.targets)
+
+        self.group_examples()
+
+    def group_examples(self):
+        """
+        To ease the accessibility of data based on the class, we will use `group_examples` to group
+        examples based on class. The data classes have already been mapped to numeric values and
+        so are the target outputs for each training input
+
+        Every key in `grouped_examples` corresponds to a class in CIFAR10 dataset. For every key in
+        `grouped_examples`, every value will conform to all of the indices for the CIFAR10
+        dataset examples that correspond to that key.
+        """
+
+        # get the targets from CIFAR10 dataset
+        np_arr = np.array(self.dataset.targets.clone())
+
+        # group examples based on class
+        self.grouped_examples = {}
+        for i in range(0, 10):
+            self.grouped_examples[i] = np.where((np_arr == i))[0]
+
+    def __len__(self):
+        return self.data.shape[0]
+
+    def __getitem__(self, index):
+        """
+        For every sample in the batch we select 3 images. First one is the anchor image
+        which is the image obtained from the current index. We also obtain the label of
+        anchor image.
+
+        Now we select two random images, one belonging to the same class as that of the
+        anchor image (named as positive_image) and the other belonging to a different class
+        than that of the anchor image (named as negative_image). We return the anchor image,
+        positive image, negative image and anchor label.
+        """
+
+        # obtain the anchor image
+        anchor_image = self.data[index].clone().float()
```

Comment:
Do we need to copy data here ?

Reply:
I believe it might be the reason for taking access time. I will remove it and check

Reply:
Same here, unaddressed comment